File: MATLAB/step11.m

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Step 11: Cavity Flow 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

clc; clear; close;

X = 41;
Y = 41;
T = 100;

  c = 1.000;
 dt = 0.001;
 nu = 0.100;
rho = 1.000;

dx = 2 / (X - 1);
dy = 2 / (Y - 1);

x = linspace(0, 2, X);
y = linspace(0, 2, Y);

[nX, nY] = meshgrid(x, y);

b = zeros(X, Y);
p = zeros(X, Y);
u = zeros(X, Y);
v = zeros(X, Y);

% Lid-driven cavity condition
u(end, :) = 1;

for n = 1:T

    un = u;
    vn = v;

    % Source term b calculation
    for i = 2:X-1
        for j = 2:Y-1
            b(i,j) = rho * (1 / dt * ((u(i+1,j) - u(i-1,j)) / (2 * dx) + (v(i,j+1) - v(i,j-1)) / (2 * dy)) - ((u(i+1,j) - u(i-1,j)) / (2 * dx))^2 - 2 * ((u(i,j+1) - u(i,j-1)) / (2 * dy) * (v(i+1,j) - v(i-1,j)) / (2 * dx)) - ((v(i,j+1) - v(i,j-1)) / (2 * dy))^2);
        end
    end

    % Pressure Poisson equation
    for it = 1:50
        pn = p;
        for i = 2:X-1
            for j = 2:Y-1
                p(i,j) = ((pn(i+1,j) + pn(i-1,j)) * dy^2 + (pn(i,j+1) + pn(i,j-1)) * dx^2 - b(i,j) * dx^2 * dy^2) / (2 * (dx^2 + dy^2));
            end
        end

        % Pressure boundary conditions
        p(:,1) = p(:,2);  
        p(1,:) = p(2,:);   
        p(end,:) = 0;     
        p(:,end) = p(:,end-1); 
    end

  
    % Velocity update
    for i = 2:X-1
        for j = 2:Y-1
            u(i,j) = un(i,j) - un(i,j) * dt / dx * (un(i,j) - un(i-1,j)) - vn(i,j) * dt / dy * (un(i,j) - un(i,j-1)) - dt / (2 * rho * dx) * (p(i+1,j) - p(i-1,j)) + nu * dt * ((un(i+1,j) - 2 * un(i,j) + un(i-1,j)) / dx^2 + (un(i,j+1) - 2 * un(i,j) + un(i,j-1)) / dy^2);
            v(i,j) = vn(i,j) - un(i,j) * dt / dx * (vn(i,j) - vn(i-1,j)) - vn(i,j) * dt / dy * (vn(i,j) - vn(i,j-1)) - dt / (2 * rho * dy) * (p(i,j+1) - p(i,j-1)) + nu * dt * ((vn(i+1,j) - 2 * vn(i,j) + vn(i-1,j)) / dx^2 + (vn(i,j+1) - 2 * vn(i,j) + vn(i,j-1)) / dy^2);
        end
    end

    % Velocity boundary conditions
    u(1,:) = 0; u(:,1) = 0; u(:,end) = 1; u(end,:) = 0;
    v(1,:) = 0; v(:,1) = 0; v(:,end) = 0; v(end,:) = 0;

  
    hFig = figure(1);
    contourf(nX, nY, p);
    hold on;
    quiver(nX,nY,u',v',3);
    colorbar
    xlabel('X');
    ylabel('Y');
    title('Pressure and Velocity Field');

end

File: Python/step11.py

import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation

########################################################
# Step 11: Cavity Flow 
########################################################

X = 41
Y = 41
T = 500

dx = 2 / (X - 1)
dy = 2 / (Y - 1)

rho = 1
nu  = 0.1
dt  = 0.001

p, b = np.zeros((Y, X)), np.zeros((Y, X))
u, v = np.zeros((Y, X)), np.zeros((Y, X))

x = np.linspace(0, 2, X)
y = np.linspace(0, 2, Y)
nX, nY = np.meshgrid(x, y)

# Lid-driven cavity condition
u[-1, :] = 1 

# Setup the plot
fig, ax = plt.subplots()
speed   = np.sqrt(u**2 + v**2)
contour = ax.contourf(nX, nY, p, cmap='jet')

streamplot = ax.streamplot(nX, nY, u, v, color=speed, linewidth=0.5, cmap='jet')
cbar       = fig.colorbar(streamplot.lines)
ax.set_ylim([0, 2.25])

# Pressure Poisson equation
def pressure_poisson(p, dx, dy, b):
    for _ in range(50):
        pn = p.copy()
        for i in range(1, Y - 1):
            for j in range(1, X - 1):
                p[i, j] = (((pn[i, j+1] + pn[i, j-1]) * dy**2 +
                            (pn[i+1, j] + pn[i-1, j]) * dx**2) /
                           (2 * (dx**2 + dy**2)) -
                           dx**2 * dy**2 / (2 * (dx**2 + dy**2)) * b[i, j])

        # Boundary conditions
        p[:, -1], p[0, :], p[:, 0] = p[:, -2], p[1, :], p[:, 1]
        p[-1, :] = 0
    return p

def animate(n):
    global u, v, p

    for i in range(1, Y - 1):
        for j in range(1, X - 1):
            b[i, j] = (rho * (1 / dt * 
                              ((u[i, j+1] - u[i, j-1]) / (2 * dx) + 
                               (v[i+1, j] - v[i-1, j]) / (2 * dy)) -
                              ((u[i, j+1] - u[i, j-1]) / (2 * dx))**2 -
                              2 * ((u[i+1, j] - u[i-1, j]) / (2 * dy) *
                                   (v[i, j+1] - v[i, j-1]) / (2 * dx))-
                              ((v[i+1, j] - v[i-1, j]) / (2 * dy))**2))

    p = pressure_poisson(p, dx, dy, b)

    for i in range(1, Y - 1):
        for j in range(1, X - 1):
            u[i, j] = (u[i, j] -
                       u[i, j] * dt / dx * (u[i, j] - u[i, j-1]) -
                       v[i, j] * dt / dy * (u[i, j] - u[i-1, j]) -
                       dt / (2 * rho * dx) * (p[i, j+1] - p[i, j-1]) +
                       nu * (dt / dx**2 *
                             (u[i, j+1] - 2 * u[i, j] + u[i, j-1]) +
                             dt / dy**2 *
                             (u[i+1, j] - 2 * u[i, j] + u[i-1, j])))
            v[i, j] = (v[i, j] -
                       u[i, j] * dt / dx * (v[i, j] - v[i, j-1]) -
                       v[i, j] * dt / dy * (v[i, j] - v[i-1, j]) -
                       dt / (2 * rho * dy) * (p[i+1, j] - p[i-1, j]) +
                       nu * (dt / dx**2 *
                             (v[i, j+1] - 2 * v[i, j] + v[i, j-1]) +
                             dt / dy**2 *
                             (v[i+1, j] - 2 * v[i, j] + v[i-1, j])))

    # Update boundary conditions
    u[0, :]  = 0
    u[:, 0]  = 0
    v[0, :]  = 0
    v[:, 0]  = 0

    u[-1, :] = 1 
    u[:, -1] = 0
    v[-1, :] = 0
    v[:, -1] = 0

    speed = np.sqrt(u**2 + v**2)
    ax.clear()
    contour = ax.contourf(nX, nY, p, cmap='jet')
    streamplot = ax.streamplot(nX, nY, u, v, color=speed, linewidth=0.5, cmap='jet')
    
    ax.set_xlabel('X')
    ax.set_ylabel('Y')
    ax.set_xlim([0, 2])
    ax.set_ylim([0, 2.25])
    ax.set_title('Pressure and Velocity Field at Time = {}'.format(n))

    return contour, streamplot.lines

ani = animation.FuncAnimation(fig, animate, frames=T, interval=0.5)

plt.show()
